multicluster/controllers/multicluster/node_controller.go

```diff
@@ -73,65 +78,196 @@ func NewNodeReconciler(
+		return err
 	}
 
-	gw.GatewayIP = gwIP
-	gw.InternalIP = internalIP
-	if err := r.Client.Update(ctx, gw, &client.UpdateOptions{}); err != nil {
-		return ctrl.Result{}, err
+	gwList := &mcsv1alpha1.GatewayList{}
+	if err := r.Client.List(ctx, gwList, &client.ListOptions{}); err != nil {
+		return err
 	}
-	return ctrl.Result{}, nil
+	// Gateway webhook guarantees that there is at most one Gateway in the member cluster.
+	if len(gwList.Items) > 0 {
+		existingGWName := gwList.Items[0].Name
+		node := &corev1.Node{}
+		if err := r.Client.Get(ctx, types.NamespacedName{Name: existingGWName}, node); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return err
+			}
+			staleGateway := &mcsv1alpha1.Gateway{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: r.namespace,
+					Name:      existingGWName},
+			}
+			err := r.Client.Delete(ctx, staleGateway, &client.DeleteOptions{})
+			if err != nil && !apierrors.IsNotFound(err) {
+				return err
+			}
+		} else {
+			r.activeGateway = existingGWName
+		}
+	}
+	for _, n := range nodeList.Items {
+		if _, isGW := n.Annotations[common.GatewayAnnotation]; isGW {
+			r.gatewayCandidates[n.Name] = true
+		}
+	}
+	return nil
+}
+
+func (r *NodeReconciler) updateActiveGateway(ctx context.Context, newGateway *mcsv1alpha1.Gateway) error {
+	existingGW := &mcsv1alpha1.Gateway{}
+	// TODO: cache might be stale. Need to revisit here and other reconcilers to
+	// check if we can improve this with 'Owns' or other methods.
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: newGateway.Name, Namespace: r.namespace}, existingGW); err != nil {
+		return err
+	}
+	if existingGW.GatewayIP == newGateway.GatewayIP && existingGW.InternalIP == newGateway.InternalIP {
+		return nil
+	}
+	existingGW.GatewayIP = newGateway.GatewayIP
+	existingGW.InternalIP = newGateway.InternalIP
+	if err := r.Client.Update(ctx, existingGW, &client.UpdateOptions{}); err != nil {
+		return err
+	}
+	return nil
+}
+
+// recreateActiveGateway will delete the existing Gateway CR and create a new Gateway
+// from the pool of Gateway candidates.
+func (r *NodeReconciler) recreateActiveGateway(ctx context.Context, gateway *mcsv1alpha1.Gateway) error {
+	err := r.Client.Delete(ctx, gateway, &client.DeleteOptions{})
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	r.activeGateway = ""
+	// Check remaining Gateway candidates and create a new Gateway.
+	newGateway, err := r.getValidGatewayFromCandidates()
+	if err != nil {
+		return err
+	}
+	if newGateway != nil {
+		return r.createGateway(newGateway)
+	}
+	return nil
+}
+
+// getValidGatewayFromCandidates picks a valid Node from Gateway candidates and
+// creates a Gateway. It returns no error if no good Gateway candidate.
+func (r *NodeReconciler) getValidGatewayFromCandidates() (*mcsv1alpha1.Gateway, error) {
+	var activeGateway *mcsv1alpha1.Gateway
+	var internalIP, gwIP string
+	var err error
+	gatewayNode := &corev1.Node{}
+	if len(r.gatewayCandidates) == 0 {
```

Comment:
This block is not necessary and can be removed.

Reply:
Removed